File: src/com/johnlindquist/acejump/AceCanvas.java
package com.johnlindquist.acejump;

import com.intellij.openapi.util.Pair;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.*;
import java.util.List;

/**
 * Created with IntelliJ IDEA.
 * User: John
 * Date: 8/10/12
 * Time: 9:45 AM
 */
public class AceCanvas extends JComponent {
    private List<Pair<String, Point>> ballonInfos;
    private Pair<Color, Color> colorPair;
    private float lineSpacing;
    private int lineHeight;


    public void setBallonInfos(@Nullable List<Pair<String, Point>> ballonInfos) {
        this.ballonInfos = ballonInfos;
    }

    @Override
    public void paint(Graphics g) {
        super.paint(g);    //To change body of overridden methods use File | Settings | File Templates.
        if (ballonInfos == null) return;

        Font font = getFont();
        Graphics2D g2d = (Graphics2D) g;
        int fontWidth = getFontMetrics(font).stringWidth("w");
        int fontHeight = font.getSize();

        int rectMarginWidth = fontWidth / 2;
        int doubleRectMarginWidth = rectMarginWidth * 2;
        int rectHOffset = (int) (fontHeight * lineSpacing - fontHeight);
        int rectWidth = fontWidth + doubleRectMarginWidth;

        for (Pair<String, Point> ballonInfo : ballonInfos) {

            String text = ballonInfo.getFirst();
            Point originalPoint = ballonInfo.getSecond();
            Color defaultForeground = colorPair.getSecond();
            Color defaultBackground = colorPair.getFirst();


            g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);

            //a slight border for "pop"

            g2d.setColor(defaultBackground);
            g2d.drawRect(originalPoint.x - rectMarginWidth - 1, originalPoint.y - rectHOffset - 1, rectWidth + 1, lineHeight + 1);

            //the background rectangle
            g2d.setColor(defaultForeground);
            g2d.fillRect(originalPoint.x - rectMarginWidth, originalPoint.y - rectHOffset, rectWidth, lineHeight);

            //just a touch of alpha
            g2d.setComposite(AlphaComposite.getInstance(AlphaComposite.SRC_OVER, 0.85f));

            //the foreground text
            g2d.setFont(font);
            g2d.setColor(defaultBackground);
            g2d.drawString(text.toUpperCase(), originalPoint.x, originalPoint.y + fontHeight);

        }

    }

    public void setBackgroundForegroundColors(Pair<Color, Color> colorPair) {
        this.colorPair = colorPair;
    }


    public void setLineSpacing(float lineSpacing) {
        this.lineSpacing = lineSpacing;
    }

    public void setLineHeight(int lineHeight) {
        this.lineHeight = lineHeight;
    }
}

File: src/main/kotlin/com/johnlindquist/acejump/ui/AceCanvas.kt
package com.johnlindquist.acejump.ui

import com.intellij.openapi.editor.colors.EditorColorsManager
import com.intellij.openapi.editor.impl.EditorImpl
import java.awt.Font
import java.awt.Font.BOLD
import java.awt.Graphics
import java.awt.Graphics2D
import javax.swing.JComponent

class AceCanvas(val editor: EditorImpl) : JComponent() {
  var jumpLocations: Collection<JumpInfo> = arrayListOf()
  val scheme = EditorColorsManager.getInstance().globalScheme
  val fbm: FontBasedMeasurements
  var existingTags = hashSetOf<Pair<Int, Int>>()

  init {
    font = Font(scheme.editorFontName, BOLD, scheme.editorFontSize)
    fbm = FontBasedMeasurements()
  }

  inner class FontBasedMeasurements() {
    var font = getFont()!!
    val fontWidth = getFontMetrics(font).stringWidth("w")
    val fontHeight = font.size
    val lineHeight = editor.lineHeight
    val lineSpacing = scheme.lineSpacing
    val fontSpacing = fontHeight * lineSpacing
    val rectHOffset = fontSpacing - fontHeight
    val hOffset = fontHeight - fontSpacing
  }

  override fun paint(graphics: Graphics) {
    if (jumpLocations.isEmpty())
      return

    super.paint(graphics)

    val g2d = graphics as Graphics2D
    existingTags = hashSetOf<Pair<Int, Int>>()
    jumpLocations.forEach { it.paintMe(g2d, this@AceCanvas) }
  }

  fun registerTag(point: Pair<Int, Int>, tag: String) {
    (-1..(tag.length)).forEach {
      existingTags.add(Pair(point.first + it * fbm.fontWidth, point.second))
    }
  }

  fun isFree(point: Pair<Int, Int>): Boolean {
    return !existingTags.contains(point)
  }
}